File: Java8-JavaFX/src/sevenguis/timer/TimerReactFX.java
package sevenguis.timer;

import javafx.application.Application;
import javafx.event.ActionEvent;
import javafx.geometry.Insets;
import javafx.scene.Scene;
import javafx.scene.control.Button;
import javafx.scene.control.Label;
import javafx.scene.control.ProgressBar;
import javafx.scene.control.Slider;
import javafx.scene.layout.HBox;
import javafx.scene.layout.VBox;
import javafx.stage.Stage;
import org.reactfx.EventStream;
import org.reactfx.EventStreams;
import org.reactfx.StateMachine;
import org.reactfx.util.Tuples;

import java.time.Duration;

import static org.reactfx.EventStreams.*;

// https://gist.github.com/TomasMikula/1013e56be2f282416274
public class TimerReactFX extends Application {

    public void start(Stage stage) {
        ProgressBar progress = new ProgressBar();
        Label numericProgress = new Label();
        Slider slider = new Slider(1, 400, 200);
        Button reset = new Button("Reset");

        EventStream<ActionEvent> resets = EventStreams.eventsOf(reset, ActionEvent.ACTION);
        EventStream<?> ticks = EventStreams.ticks(Duration.ofMillis(100));
        EventStream<Double> elapsed = StateMachine.init(Tuples.t(0.0, slider.getValue()))
                .on(resets).transition((tup_es, r) -> Tuples.t(0.0, tup_es._2))
                .on(ticks).transition((tup_es, t) -> Tuples.t(tup_es._1 + (tup_es._1 < tup_es._2 ? 1 : 0), tup_es._2))
                .on(EventStreams.valuesOf(slider.valueProperty())).transition((tup_es, s1) -> Tuples.t(tup_es._1, s1.doubleValue()))
                .toStateStream().map(tup_es -> tup_es._1);

        combine(elapsed, valuesOf(slider.valueProperty())).map((e, s) -> e / s.doubleValue()).subscribe(progress::setProgress);
        elapsed.map(TimerReactFX::formatElapsed).subscribe(numericProgress::setText);

        VBox root = new VBox(10, new HBox(10, new Label("Elapsed Time: "), progress),
                                 numericProgress,
                                 new HBox(10, new Label("Duration: "), slider),
                                 reset);
        root.setPadding(new Insets(10));

        stage.setScene(new Scene(root));
        stage.setTitle("Timer ReactFX");
        stage.show();
    }

    private static String formatElapsed(double elapsed) {
        int seconds = (int) Math.floor(elapsed / 10.0);
        int dezipart = (int) elapsed % 10;
        return seconds + "." + dezipart + "s";
    }

    public static void main(String[] args) {
        launch(args);
    }
}

File: Scala-ScalaFX/src/main/scala/sevenguis/timer/TimerReactFX.scala
package sevenguis.timer

import java.time.Duration

import org.reactfx.{StateMachine, EventStreams, EventStream}

import scalafx.application.JFXApp
import scalafx.application.JFXApp.PrimaryStage
import scalafx.scene.Scene
import scalafx.scene.control._
import scalafx.scene.layout.{VBox, HBox}
import scalafx.geometry.Insets

import org.reactfx.EventStreams._
import sevenguis.ReactFXIntegration._
import sevenguis.Scala2Java8._

// https://gist.github.com/TomasMikula/1013e56be2f282416274
object TimerReactFX extends JFXApp {
  val progress = new ProgressBar()
  val numericProgress = new Label()
  val slider = new Slider(1, 400, 200)
  val reset = new Button("Reset")

  type ES = (Double, Double)
  val onReset: Function2[ES, Any, ES] = {case ((_, s), _) => (0.0, s)}
  val onTick: Function2[ES, Any, ES] = {case ((e, s), _) => (e + (if (e < s) 1 else 0), s)}
  val onSlide: Function2[ES, Number, ES] = {case ((e, _), s1) => (e, s1.doubleValue())}
  val pickElapsed: Function1[ES, Double] = {case (e, _) => e}

  val resets = reset.actions
  val ticks: EventStream[_] = EventStreams.ticks(Duration.ofMillis(100))
  val elapsed: EventStream[Double] = StateMachine.init((0.0, slider.value()))
          .on(resets).transition(onReset)
          .on(ticks).transition(onTick)
          .on(slider.value).transition(onSlide)
          .toStateStream.map(pickElapsed)

  progress.progress |= combine(elapsed, slider.value).map(((e, s) => e / s.doubleValue()): ((Double, Number) => Number))
  numericProgress.text |= elapsed.map((e => formatElapsed(e)): (Double => String))

  stage = new PrimaryStage {
    title = "Timer"
    scene = new Scene {
      content = new VBox(10) {
        padding = Insets(10)
        content =
          Seq(new HBox(10) { content = Seq(new Label("Elapsed Time: "), progress) },
              numericProgress,
              new HBox(10) { content = Seq(new Label("Duration: "), slider) },
              reset)
      }
    }
  }

  def formatElapsed(elapsed: Double): String = {
    val seconds = (elapsed / 10.0).floor
    val dezipart = elapsed % 10
    seconds.toInt + "." + dezipart.toInt + "s"
  }
}
